src/System.Net.Http/src/System/Net/Http/Managed/AuthenticationHelper.Digest.cs

```diff
@@ -0,0 +1,374 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Net.Http.Headers;
+using System.Security.Cryptography;
+using System.Text;
+
+namespace System.Net.Http
+{
+    internal partial class AuthenticationHelper
+    {
+        public const string Digest = "Digest";
+
+        // Define digest constants
+        private const string Qop = "qop";
+        private const string Auth = "auth";
+        private const string AuthInt = "auth-int";
+        private const string Nonce = "nonce";
+        private const string NC = "nc";
+        private const string Realm = "realm";
+        private const string UserHash = "userhash";
+        private const string Username = "username";
+        private const string Algorithm = "algorithm";
+        private const string Uri = "uri";
+        private const string Sha256 = "SHA-256";
+        private const string Md5 = "MD5";
+        private const string Sha256Sess = "SHA-256-sess";
+        private const string CNonce = "cnonce";
+        private const string Opaque = "opaque";
+        private const string Response = "response";
+        private const string Colon = ":";
+
+        // Define alphanumeric characters for cnonce
+        // 48='0', 65='A', 97='a'
+        private static int[] AlphaNumChooser = new int[] { 48, 65, 97 };
+
+        public static bool TrySetDigestAuthToken(HttpRequestMessage request, ICredentials credentials, DigestResponse digestResponse)
+        {
+            NetworkCredential credential = credentials.GetCredential(request.RequestUri, Digest);
+            if (credential == null)
+            {
+                return false;
+            }
+
+            try
+            {
+                request.Headers.Authorization = new AuthenticationHeaderValue(Digest, GetDigestTokenForCredential(credential, request, digestResponse));
+                return true;
+            }
+            catch
+            {
+                // Return false in case of any digest header calculation errors.
```

Comment:
The exceptions that might arise is because of KeyNotFoundException from the parameter dictionary. If any server doesn't implement digest according to the rfc, and skips any mandatory fields like nonce, etc, for which I don't do a ContainsKey check before accessing it. In that case we want the authentication to fail, since the server challenge is malformed. So i'm catching those exceptions here and return false. To make it exception free, i can make GetDigestTokenForCredential return null string, and check for null and return false. Which is better?